crates/accelerate/src/quantum_circuit/circuit_instruction.rs

```diff
@@ -0,0 +1,256 @@
+                        ))
+                    }
+                }
+            }
+        }
+
+        Ok(CircuitInstruction {
+            operation,
+            qubits: as_tuple(py, qubits)?,
+            clbits: as_tuple(py, clbits)?,
+        })
+    }
+
+    /// Returns a shallow copy.
+    ///
+    /// Returns:
+    ///     CircuitInstruction: The shallow copy.
+    pub fn copy(&self) -> Self {
+        self.clone()
+    }
+
+    /// Creates a shallow copy with the given fields replaced.
+    ///
+    /// Returns:
+    ///     CircuitInstruction: A new instance with the given fields replaced.
+    pub fn replace(
+        &self,
+        py: Python<'_>,
+        operation: Option<PyObject>,
+        qubits: Option<&PyAny>,
+        clbits: Option<&PyAny>,
+    ) -> PyResult<Self> {
+        CircuitInstruction::new(
+            py,
+            operation.unwrap_or_else(|| self.operation.clone_ref(py)),
+            Some(qubits.unwrap_or_else(|| self.qubits.as_ref(py))),
+            Some(clbits.unwrap_or_else(|| self.clbits.as_ref(py))),
+        )
+    }
+
+    fn __getstate__(&self, py: Python<'_>) -> PyObject {
+        (
+            self.operation.as_ref(py),
+            self.qubits.as_ref(py),
+            self.clbits.as_ref(py),
+        )
+            .into_py(py)
+    }
+
+    fn __setstate__(&mut self, _py: Python<'_>, state: &PyTuple) -> PyResult<()> {
+        self.operation = state.get_item(0)?.extract()?;
+        self.qubits = state.get_item(1)?.extract()?;
+        self.clbits = state.get_item(2)?.extract()?;
+        Ok(())
+    }
+
+    pub fn __getnewargs__(&self, py: Python<'_>) -> PyResult<PyObject> {
+        Ok((
+            self.operation.as_ref(py),
+            self.qubits.as_ref(py),
+            self.clbits.as_ref(py),
+        )
+            .into_py(py))
+    }
+
+    pub fn __repr__(self_: &PyCell<Self>, py: Python<'_>) -> PyResult<String> {
+        let type_name = self_.get_type().name()?;
+        let r = self_.try_borrow()?;
+        Ok(format!(
+            "{}(\
+            operation={}\
+            , qubits={}\
+            , clbits={}\
+            )",
+            type_name,
+            r.operation.as_ref(py).repr()?,
+            r.qubits.as_ref(py).repr()?,
+            r.clbits.as_ref(py).repr()?
+        ))
+    }
+
+    // Legacy tuple-like interface support.
+    //
+    // For a best attempt at API compatibility during the transition to using this new class, we need
+    // the interface to behave exactly like the old 3-tuple `(inst, qargs, cargs)` if it's treated
+    // like that via unpacking or similar.  That means that the `parameters` field is completely
+    // absent, and the qubits and clbits must be converted to lists.
+    pub fn _legacy_format(&self, py: Python<'_>) -> PyObject {
+        // Note: previously, we converted qubits and clbits to list,
+        //       but this appears no longer necessary.
```

Comment:
This is likely only because I squashed all use of the legacy format from Terra. As long as we're supporting the legacy format, we should convert to list, because there absolutely were lots of places that assumed it was a list, which means it's likely to be the case in user code as well.

Reply:
I hadn't thought about user code, good point.

Added conversion to list for legacy format bit lists in dad343051472dc9ff4d2662a27e8de1c2986abb1.